Sources/TuistGenerator/Generator/DescriptorGenerator.swift

```diff
@@ -0,0 +1,85 @@
+import Basic
+import Foundation
+import TuistCore
+import TuistSupport
+import XcodeProj
+
+public struct GeneratedSchemeDescriptor {
+    public var scheme: XCScheme
+    public var shared: Bool
+}
+
+public enum GeneratedSideEffect {
+    case file(GeneratedFile)
```

Comment:
What about following an API similar to [Ansible's](https://docs.ansible.com/ansible/latest/modules/file_module.html#parameter-state). Rather than describing an imperative action, I'd describe a state:
- The file x needs to exist with this content.
- The file x should be absent.
- The directory y should exist with the following permissions.

Then it's the component that resolves the side effects the one that figures out what's necessary to get the system to that state.